src/profile-logic/js-tracer.js

```diff
@@ -0,0 +1,340 @@
+      const otherEnd = timingRow.end[timingRow.length - 1];
+      if (otherEnd === undefined || otherEnd <= start) {
+        timingRow.start.push(start);
+        timingRow.end.push(start + duration);
+        timingRow.label.push(displayName);
+        timingRow.index.push(tracerEventIndex);
+        timingRow.length++;
+        break;
+      }
+    }
+  }
+
+  return jsTracerTiming;
+}
+
+/**
+ * Determine the self time for JS Tracer events. This generates a row of timing
+ * information for each type event. This function walks the stack structure of JS Tracer
+ * events, and determines what's actually executing at a given time.
+ *
+ * The data visually would look something like this:
+ *
+ * A: [----]          [-----]
+ * B:       [-----]
+ * C:              [-]       [--]
+ * D:                            [----]
+ *
+ * Where "A" would be the name of the event, and the boxes in that row would be the self
+ * time. Each row is stored as a JsTracerTiming.
+ */
+export function getJsTracerLeafTiming({
+  events: tracerEvents,
+  stringTable,
+}: JsTracerTable): JsTracerTiming[] {
+  // Each event type will have it's own timing information, later collapse these into
+  // a single array.
+  const jsTracerTimingMap: Map<string, JsTracerTiming> = new Map();
+  const isUrl = stringTable._array.map(string => /:\/\//.test(string));
+
+  function reportSelfTime(
+    tracerEventIndex: number,
+    start: Microseconds,
+    end: Microseconds
+  ): void {
+    if (start === end) {
+      // If this self time is 0, do not report it.
+      return;
+    }
+    const stringIndex = tracerEvents.events[tracerEventIndex];
+    const displayName = stringTable.getString(stringIndex);
+    const rowName = isUrl[stringIndex] ? 'Script' : displayName;
+    let timingRow = jsTracerTimingMap.get(rowName);
+    if (timingRow === undefined) {
+      timingRow = {
+        start: [],
+        end: [],
+        index: [],
+        label: [],
+        name: rowName,
+        length: 0,
+      };
+      jsTracerTimingMap.set(rowName, timingRow);
+    }
+    // Convert the timing to milliseconds.
+    const division = 1000;
+    if (end < start) {
+      throw new Error('end is less than the start');
+    }
+    timingRow.start.push(start / division);
+    timingRow.end.push(end / division);
+    timingRow.label.push(displayName);
+    timingRow.index.push(tracerEventIndex);
+    timingRow.length++;
+
+    if (timingRow.length > 1) {
+      const currStart = timingRow.start[timingRow.length - 1];
+      const currEnd = timingRow.end[timingRow.length - 1];
+      const prevEnd = timingRow.end[timingRow.length - 2];
+      if (currEnd < currStart) {
+        throw new Error(
+          `currEnd < currStart "${displayName} - ${currEnd} < ${currStart}"`
+        );
+      }
+      if (currStart < prevEnd) {
+        throw new Error(
+          `currStart < prevEnd "${displayName} - ${currStart} < ${prevEnd}"`
+        );
+      }
```

Comment:
This is the invariant that tests that the information is truly a tree data structure. I believe that this check should exist, and should be this harsh, as it means we are broken and can't rely on the information. Plus, I want to know if we fail if it's our fault, or the data's fault.

@dpalmeiro Do you agree that it's OK to throw a client-side error (essentially break the tools) if the data source is wrong? The other choice would be to gracefully handle the error and display a nice user message. This would add more UI code and complexity, but it is something we could do. My preference would be to fail here. If we want handle this without throwing an error, then I would prefer to add this as a follow-up.

Reply:
Oh woops, this would be a client-only error I THINK. But this code is really complex to understand, and I would still prefer to throw errors aggressively.

Also, the checks might be implied by each other, but I'd prefer to be extra careful with this code, as it has many possible failure states.

Reply:
I agree with the later comment that we should have an error boundary in the React component to display an error. I filed it as a follow-up, and will resolve this comment: https://github.com/devtools-html/perf.html/issues/1503

Reply:
Yes, I think it's good to have these check, especially as more events will be added in the future.  One thing to note, however, is that these checks fail when I try to truncate the double precision values into an integer.  I think we need a "<=" test somewhere if we decide to do that in the future (e.g. when https://bugzilla.mozilla.org/show_bug.cgi?id=1503959 is implemented, everything will probably be an integer).

Reply:
Throwing errors is a very scary option at the moment, my concern is that the user might have a properly functioning profile view until they move to the "jstracer" panel, which throws, and you know what happens in React when this throws at render.

This is much less scary with error boundaries. So I'm fine keeping these errors in if we implement error boundaries sooner rather than later.

Reply:
Unresolving mainly for @dpalmeiro's comment above. I think we should support the future especially when we know what the future will be (lucky we are :) ).

Reply:
My preference here would be to follow-up on error boundaries as this is an experimental feature with very low usage. I would prefer to get it into the hands of users to validate the feature sooner rather than later, so things breaking seems like an acceptable risk to me. This isn't going out to our general population.

Please let me know if you want to block this PR on #1503 landing first. I don't want to do the error boundary work in this PR, as it's quite large already. So I will do the work in a separate PR.

Reply:
Error Boundaries should happen in a separate PR for sure.

I guess it depends on whether we expect throwing errors often. If this is really unexpected, then I guess we can land it without the error boundaries.